### ISSUE TITLE
Add nightly build for detection of flaky tests

### ISSUE THREAD

**Issue by osmaczko** [status-im/status-go#4264]
# Problem

Lately, we have been grappling with a significant number of unstable tests. Currently, pinpointing these erratic tests requires manual effort, which involves observing which tests fail upon re-execution. This approach is far from ideal.

Implementing a nightly build process that executes each test multiple times and generates an accessible report would greatly assist in detecting unstable tests. Additionally, this could help in tracing back to commits that potentially increased the instability of existing tests.

**Comment by osmaczko** [comment]
@jakubgs could we arrange for the infra team to implement this in the near future, please?

CC: @cammellos

**Comment by yakimant** [comment]
@osmaczko, first idea from my mind, can we actually split unit and integration tests?
For example separate makefile tasks. What do you think?

Unit tests should be always stable and running predictable time.

**Comment by osmaczko** [comment]
> @osmaczko, first idea from my mind, can we actually split unit and integration tests? For example separate makefile tasks. What do you think?
> 
> Unit tests should be always stable and running predictable time.

Yes, that would make sense, although I don't believe it should be a prerequisite for this task. We don't have a clear separation in the code right now, and it would require some extra work to identify unit tests versus integration tests. A good indicator would be the execution time; unit tests should theoretically be the fastest. I would suggest creating a separate task to identify and split the tests.

**Comment by yakimant** [comment]
Yes, that would be great. But I agree, it's not a prerequisite.

Linking just for the reference:
https://github.com/status-im/status-go/issues/3918

**Comment by yakimant** [comment]
@osmaczko, @cammellos, I need an advice.

Currently we fail test run early:
- Using `-failfast` option
- And failing jenkins job on first failing package

Fo nightly runs we obviously don't need those.

The question is - do you want not to fail early on PR test runs too?
As a benefit - developer gets the list of all failing tests in all packages.
As a downside - it will run longer.

What do you think?

**Comment by yakimant** [comment]
Here is first attempt to run tests fast and not to fail early for nightly runs:
- Old UI: https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-4295/6/testReport/
- Blue Ocean: https://ci.status.im/blue/organizations/jenkins/status-go%2Fprs%2Ftests/detail/PR-4295/6/tests/

I need to check, how jenkins handles duplicates in junit xml file (both fail, one fail, none fail).
But so far looks good.

**Comment by yakimant** [comment]
2 runs take 30 min, but I guess it can take up to 1 hour in some cases.

**Comment by cammellos** [comment]
> @osmaczko, @cammellos, I need an advice.
> 
> Currently we fail test run early:
> 
> * Using `-failfast` option
> * And failing jenkins job on first failing package
> 
> Fo nightly runs we obviously don't need those.
> 
> The question is - do you want not to fail early on PR test runs too? As a benefit - developer gets the list of all failing tests in all packages. As a downside - it will run longer.
> 
> What do you think?

I think they should failfast on PRs, most of the times you want to fix the first error you see and take it from there, rather than having to wait, but no strong opinions on my side

**Comment by osmaczko** [comment]
> > @osmaczko, @cammellos, I need an advice.
> > Currently we fail test run early:
> > 
> > * Using `-failfast` option
> > * And failing jenkins job on first failing package
> > 
> > Fo nightly runs we obviously don't need those.
> > The question is - do you want not to fail early on PR test runs too? As a benefit - developer gets the list of all failing tests in all packages. As a downside - it will run longer.
> > What do you think?
> 
> I think they should failfast on PRs, most of the times you want to fix the first error you see and take it from there, rather than having to wait, but no strong opinions on my side

I don't have a strong opinion on that. Failfast provides faster feedback, which is especially useful nowadays with the prevalence of flaky tests that often require multiple runs.

**Comment by yakimant** [comment]
Thanks! Will keep 'failfast' for now, until unit tests are split from all tests.

**Comment by yakimant** [comment]
Update on `go test` / `go-junit-report` / `jenkins` integration through junit xml files:

Jenkins treat such duplications as separate tests:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="30" failures="4">
	<testsuite name="github.com/status-im/status-go/abi-spec" tests="30" failures="3" errors="0" id="0" hostname="yakimant-macbook-air-2.home" time="0.301" timestamp="2023-11-16T12:21:50+01:00">
		<testcase name="TestEncodeTransfer" classname="github.com/status-im/status-go/abi-spec" time="0.000">
			<failure message="Failed"><![CDATA[    core_test.go:32: 
        	Error Trace:	/Users/status/work/status-go/abi-spec/core_test.go:32
        	Error:      	Not equal: 
        	            	expected: true
        	            	actual  : false
        	Test:       	TestEncodeTransfer]]></failure>
		</testcase>
		<testcase name="TestEncode" classname="github.com/status-im/status-go/abi-spec" time="0.000">
			<failure message="Failed"><![CDATA[    core_test.go:43: 
        	Error Trace:	/Users/status/work/status-go/abi-spec/core_test.go:43
        	Error:      	Not equal: 
        	            	expected: true
        	            	actual  : false
        	Test:       	TestEncode]]></failure>
		</testcase>
		<testcase name="TestEncodeTransfer" classname="github.com/status-im/status-go/abi-spec" time="0.000">
			<failure message="Failed"><![CDATA[    core_test.go:32: 
        	Error Trace:	/Users/status/work/status-go/abi-spec/core_test.go:32
        	Error:      	Not equal: 
        	            	expected: true
        	            	actual  : false
        	Test:       	TestEncodeTransfer]]></failure>
		</testcase>
		<testcase name="TestEncode" classname="github.com/status-im/status-go/abi-spec" time="0.000"></testcase>
	</testsuite>
</testsuites>
```
2 tests are failing (1 constantly, 1 randomly).

Here is how it will look lik in the UI.
Classic:
![Screenshot 2023-11-16 at 12 55 33](https://github.com/status-im/status-go/assets/1529006/f59606f4-2b07-4ab4-8217-89ed5c83441b)
Blue Ocean:
![Screenshot 2023-11-16 at 12 54 34](https://github.com/status-im/status-go/assets/1529006/084c0fab-50f0-4b6e-8940-269528281de2)

**Comment by yakimant** [comment]
@osmaczko, @cammellos, is it Ok?

I can research on options:
1. multirun support for junit/jenkins integration
2. show failing test once (no matter how many times did it fail)

Other than that nightly runs is almost ready.

**Comment by osmaczko** [comment]
> @osmaczko, @cammellos, is it Ok?
> 
> I can research on options:
> 
> 1. multirun support for junit/jenkins integration
> 2. show failing test once (no matter how many times did it fail)
> 
> Other than that nightly runs is almost ready.

Would it be possible to generate a report in the form of a table? This would allow us to quickly identify which tests are the most unstable and need immediate attention.

| test name | success rate |
| - | - |
| A | 100% |
| B | 80% (flaky) |
| C | 40% (super flaky) |

**Comment by yakimant** [comment]
@osmaczko it's deffinetly possible, but I don't see any ready solution.
It will require to make an xml/json paring script.

Let's have nightlies running with reports wvailable and I will have a report, that you suggested in a next step.

**Comment by osmaczko** [comment]
> @osmaczko it's deffinetly possible, but I don't see any ready solution. It will require to make an xml/json paring script.
> 
> Let's have nightlies running with reports wvailable and I will have a report, that you suggested in a next step.

Great, good to me, thank you :pray:

**Comment by yakimant** [comment]
@osmaczko, PR is merged and job will run nightly:
https://ci.status.im/job/status-go/job/tests-nightly/

It will run all the tests 10 times.
It provides stats in the end of the log, for example:
```
---
Failing tests stats
(test name: failure rate)
---
TestMessengerCommunitiesSuite: 100.0%
TestMessengerCommunitiesSuite/TestBanUser: 100.0%
TestMessengerCommunitiesSuite/TestCancelRequestAccess: 100.0%
TestMessengerCommunitiesSuite/TestCheckCommunitiesToUnmute: 100.0%
TestMessengerCommunitiesSuite/TestCommunityBanUserRequestToJoin: 100.0%
TestMessengerCommunitiesSuite/TestCommunityContactCodeAdvertisement: 100.0%
TestMessengerCommunitiesSuite/TestCommunityNotInDB: 100.0%
TestMessengerCommunitiesSuite/TestCommunityRekeyAfterBan: 100.0%
TestMessengerCommunitiesSuite/TestCommunityRekeyAfterBanDisableCompatibility: 100.0%
TestMessengerCommunitiesSuite/TestCreateCommunity: 100.0%
TestMessengerCommunitiesSuite/TestCreateCommunity_WithoutDefaultChannel: 100.0%
TestMessengerCommunitiesSuite/TestDeclineAccess: 100.0%
TestMessengerCommunitiesSuite/TestDeletePendingRequestAccess: 100.0%
TestMessengerCommunitiesSuite/TestDeletePendingRequestAccessWithDeclinedState: 100.0%
TestMessengerCommunitiesSuite/TestExtractDiscordChannelsAndCategories: 100.0%
TestMessengerCommunitiesSuite/TestExtractDiscordChannelsAndCategories_WithErrors: 100.0%
TestMessengerCommunitiesSuite/TestGetCommunityIdFromKey: 100.0%
TestMessengerCommunitiesSuite/TestHandleImport: 100.0%
TestMessengerCommunitiesSuite/TestImportCommunity: 100.0%
TestMessengerCommunitiesSuite/TestJoinCommunity: 100.0%
TestMessengerCommunitiesSuite/TestLeaveAndRejoinCommunity: 100.0%
TestMessengerCommunitiesSuite/TestMuteAllCommunityChats: 100.0%
TestMessengerCommunitiesSuite/TestPinMessageInCommunityChat: 100.0%
TestMessengerCommunitiesSuite/TestPostToCommunityChat: 100.0%
TestMessengerCommunitiesSuite/TestRemovePrivateKey: 100.0%
TestMessengerCommunitiesSuite/TestRequestAccess: 100.0%
TestMessengerCommunitiesSuite/TestRequestAccessAgain: 100.0%
TestMessengerCommunitiesSuite/TestRetrieveBigCommunity: 100.0%
TestMessengerCommunitiesSuite/TestRetrieveCommunity: 100.0%
TestMessengerCommunitiesSuite/TestRolesAfterImportCommunity: 100.0%
TestMessengerCommunitiesSuite/TestSetMutePropertyOnChatsByCategory: 100.0%
TestMessengerCommunitiesSuite/TestShareCommunity: 100.0%
TestMessengerCommunitiesSuite/TestShareCommunityWithPreviousMember: 100.0%
TestMessengerCommunitiesSuite/TestStartCommunityRekeyLoop: 100.0%
TestLoggerWithJSONHexEncoder: 50.0%
TestLoggerWithConsoleHexEncoder: 50.0%
TestNTPTimeSource: 50.0%
TestNTPTimeSource/SameResponse: 50.0%
TestNTPTimeSource/Median: 50.0%
TestNTPTimeSource/EvenMedian: 50.0%
TestNTPTimeSource/Error: 50.0%
TestNTPTimeSource/MultiError: 50.0%
TestNTPTimeSource/TolerableError: 50.0%
TestNTPTimeSource/NonTolerableError: 50.0%
TestNTPTimeSource/AllFailed: 50.0%
TestNTPTimeSource/HalfTolerable: 50.0%
TestClosingsqlDB: 50.0%
TestMailServerPostgresDBSuite: 50.0%
TestMailServerPostgresDBSuite/TestPostgresDB_BuildIteratorWithBloomFilter: 50.0%
TestMailServerPostgresDBSuite/TestPostgresDB_BuildIteratorWithTopic: 50.0%
TestSimulationBloomFilter: 50.0%
TestCreateAccountAndLogin: 50.0%
TestProfilingCPU: 50.0%
TestProfilingMem: 50.0%
```

I will watch the job for the next days to see if everthing is good.

**Comment by osmaczko** [comment]
@yakimant perfect, thank you!

**Comment by osmaczko** [comment]
@yakimant seems something is wrong with https://ci.status.im/job/status-go/job/tests-nightly/, could you please take a look?

**Comment by osmaczko** [comment]
@yakimant tests on latest develop are failing on CI with error: "No test result found". See: https://ci.status.im/blue/organizations/jenkins/status-go%2Fprs%2Ftests/detail/PR-4348/3/tests

I checked and it is caused by: https://github.com/status-im/status-go/pull/4295

Could you please take a look?

PS. I am wondering how did you manage to merge the PR with failing tests :thinking: 
![image](https://github.com/status-im/status-go/assets/33099791/11ec013e-fb2c-494b-96ee-250deffe188b)

**Comment by yakimant** [comment]
Small update on the issue:
- https://github.com/status-im/status-go/pull/4349 instead of printing to the build logs
- I fixed one issue with nightlies, it was trying to run on a wrong branch
- "No test result found" issue is real.
It happens, when `go test` doesn't fail the tests nicely.
For example when `panic: test timed out after 10m0s`.
`go test` fails in this case with non-zero exit code.
`go-junit-report` doesn't know if tests failed or succeed and gives an error.
In Jenkins all the tests from suite are marked as failed, because non has test results.
- `-count 1` could potentially interfere in tests cacing results:
> When 'go test' runs in package list mode, 'go test' caches successful package test results to avoid unnecessary repeated running of tests. To disable test caching, use any test flag or argument other than the cacheable flags. The idiomatic way to disable test caching explicitly is to use -count=1.

This could potentially:
1. run more on CI
2. run the unstable test again, but failing it this time

**Comment by yakimant** [comment]
One issue was discovered yesterday (kudos to @osmaczko raising it).
I removed `-timeout 30m` flag, which I though is redundant per package and we need to have per job timeout.
But it appeared, that instead of being unlimited - it became 10m default, which is not enough for protocol tests.
I reverted and added `-timeout 30m`, but now nightlies are failing tests with: `panic: test timed out after 30m0s`.

I will make a PR to calculate "per package" timeout based on runs count.

**Comment by yakimant** [comment]
Last nightly was good:
- No more timeouts
- 10 runs took 2.5h, we can increase if needed
- Both Classic and Blue Ocean don't agregate test failures by test name, so you see a lot of duplicate failed tests
- One useful thing, that both show the newly failed tests, compared to the previous nightly run:
![Screenshot 2023-11-23 at 10 39 13](https://github.com/status-im/status-go/assets/1529006/abcd9ed3-702d-46ec-99b1-4a3b14183038)
![Screenshot 2023-11-23 at 10 39 35](https://github.com/status-im/status-go/assets/1529006/ed8ff7f4-93b3-4dcc-8edc-26cb3a1c40a9)
- Another useful thing is "Test Result Trend" in Classical UI. So you can see if we are improving or not:
![Screenshot 2023-11-23 at 10 41 22](https://github.com/status-im/status-go/assets/1529006/9db1bf54-3d79-4205-b6cc-a4b3539c00ff)
- To see test stats you need to open a job and click on "Reports" on the left side pane. You will see this:
![Screenshot 2023-11-23 at 10 43 09](https://github.com/status-im/status-go/assets/1529006/a5ac038f-7432-40ce-a301-15bb828bfbea)
![Screenshot 2023-11-23 at 10 43 26](https://github.com/status-im/status-go/assets/1529006/c27f512b-72cd-47ad-a295-a0de43181276)


Find the job here:
https://ci.status.im/job/status-go/job/tests-nightly/

**Comment by yakimant** [comment]
Several issues on Jenkins to improve deduplication of multiple runs:
- https://issues.jenkins.io/browse/JENKINS-41710
- https://issues.jenkins.io/browse/JENKINS-2184
- https://issues.jenkins.io/browse/JENKINS-27395

(Teamcity for example does it)

**Comment by osmaczko** [comment]
@yakimant that is more than I have expected, thank you :pray: 

> 10 runs took 2.5h, we can increase if needed

If you could increase it to 20 runs it would be perfect :+1:

**Comment by osmaczko** [comment]
I just took a look at the report https://ci.status.im/job/status-go/job/tests-nightly/37/Reports/, it says:

`TestMessengerCommunitiesSignersSuite/TestAutoAcceptOnOwnershipChangeRequestRequired: 20.0%`

but it is listed only once in https://ci.status.im/job/status-go/job/tests-nightly/37/testReport/. Shouldn't it be listed two times (2/10 == 20% failure rate)?

**Comment by yakimant** [comment]
Hm, that's interesting - it's shown in test results 5 times (1 failing).
So it's either not enough runs (5 instead of 10) or xml file is not imported properly.

I can add smth like: `20.0% (1 of 5 failed)`

**Comment by yakimant** [comment]
- increase runs to 20
- add more info to report

in https://github.com/status-im/status-go/pull/4378

**Comment by yakimant** [comment]
Updates weekend runs:
1. "10 runs" build contained less in junit xml output (5 or 6 runs). Either remaining didn't run or `go unit` output wasn't parsed properly by `go-junit-report` tool
2. Another run had an issue parsing junit xml produced by `go-junit-report` for test stats: `xml.etree.ElementTree.ParseError: not well-formed (invalid token)`
3. "20 runs" took 3:39h compared to "10 runs" ~2:45h. I guess it's Ok.
4. Todays run is stuck on `github.com/status-im/status-go/api` testing. I will stop it now. I will add a job timeout - 5h.

**Comment by yakimant** [comment]
Timeout was 8h actually, I've set to 5h.

**Comment by yakimant** [comment]
Again stuck on `api` tests:
```
2023-11-29T02:53:00.101Z] Testing: github.com/status-im/status-go/api
[2023-11-29T07:48:07.558Z] Sending interrupt signal to process
```

We don't have anything in stdout and we dont' publish log until https://github.com/status-im/status-go/pull/4384 is merged

**Comment by yakimant** [comment]
Nice, we have log & xml files uploaded, so we can invedtigate the following 3 issues:
1. Sometimes stuck indefinetly on `github.com/status-im/status-go/api`
2. Test stats failing to parse xml files with `xml.etree.ElementTree.ParseError: not well-formed (invalid token): line 911, column 25`
3. Amount of some test runs is less, than expected. Not parsed or not run?

**Comment by yakimant** [comment]
As for first issue, @osmaczko, can you please have a look? - this is issue number 1 in my list above
`api/test.log` log file should help.

You can see the place, where it was stuck in the end of the log, for example:
```
WARN [12-10|03:15:14.012|go.uber.org/zap/sugar.go:173]                                                      error bumping delivery tag: no decaying tag registered for topic /waku/2/default-waku/proto 
FAIL	github.com/status-im/status-go/api	17689.821s
```

**Comment by yakimant** [comment]
Pinned build for issue 2 investigation:
https://ci.status.im/job/status-go/job/tests-nightly/57/
(so the artifacts are not cleaned up)

Also issue 3, because protocol tests are run only 5 times, not 20:
https://ci.status.im/job/status-go/job/tests-nightly/57/testReport/github/com_status-im_status-go_protocol/

**Comment by yakimant** [comment]
As of issue with parsing xml files, I found a problematic place - number 2 in my list above

![Screenshot 2023-12-25 at 14 57 04](https://github.com/status-im/status-go/assets/1529006/6d3cbc08-fb62-4043-a6ad-d816f5e03907)
https://ci.status.im/job/status-go/job/tests-nightly/72/artifact/node/report.xml

![Screenshot 2023-12-25 at 14 59 04](https://github.com/status-im/status-go/assets/1529006/8f6180f0-3589-473c-acad-4a684dc6772f)
https://ci.status.im/job/status-go/job/tests-nightly/72/artifact/node/test.log

Smth got broken with log output:
- See log level/date/file placement and formatting
- Also Unicode special symbols are there, which supposed to be fixed by my PR previously: https://github.com/status-im/status-go/blob/843bae56597ecf3811d724504bd0216b867979f7/logutils/override.go#L51-L53

In this build #72 both test_stats script and Jenkins test results import failed.
I will check another case, where only script failed, but test results were imported.

**Comment by yakimant** [comment]
Issue with less test runs, than expected - number 3 in my list above

XML shows 5  times:
```
❯ grep "TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions\"" protocol/report.xml
                <testcase name="TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions" classname="github.com/status-im/status-go/protocol" time="12.830">
                <testcase name="TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions" classname="github.com/status-im/status-go/protocol" time="22.990">
                <testcase name="TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions" classname="github.com/status-im/status-go/protocol" time="13.100">
                <testcase name="TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions" classname="github.com/status-im/status-go/protocol" time="13.060">
                <testcase name="TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions" classname="github.com/status-im/status-go/protocol" time="13.000">
```
https://ci.status.im/job/status-go/job/tests-nightly/72/artifact/protocol/report.xml

Log also shows 5 times:
```
❯ grep "=== RUN   TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions" protocol/test.log
=== RUN   TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions
=== RUN   TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions
=== RUN   TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions
=== RUN   TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions
=== RUN   TestMessengerCommunitiesTokenPermissionsSuite/TestBecomeMemberPermissions
```
https://ci.status.im/job/status-go/job/tests-nightly/72/artifact/protocol/test.log

This has happened with all protocol tests, they were running 4-5 times:
```
❯ grep "=== RUN   [a-zA-Z]*$" protocol/test.log | sort | uniq -c | head
   5 === RUN   TestActivityCenterPersistence
   5 === RUN   TestAdminCommunityEventsSuite
   5 === RUN   TestBrowserSuite
   5 === RUN   TestChatSuite
   4 === RUN   TestConfirmations
   4 === RUN   TestConfirmationsAtLeastOne
   5 === RUN   TestContactAcceptContactRequest
   5 === RUN   TestContactContactRequestAccepted
   5 === RUN   TestContactContactRequestPropagatedStateReceivedOutOfDateLocalStateOnOurSide
   5 === RUN   TestContactContactRequestPropagatedStateReceivedOutOfDateLocalStateOnTheirSide
```

Looks like `go test` crashed in this case:
```
=== RUN   TestMessengerProfilePictureHandlerSuite/TestE2eSendingReceivingProfilePicture
2023-12-24T05:37:30.457Z	DEBUG	Waku	waku/waku.go:125	starting waku with config	{"config": {"MaxMessageSize":1048576,"MinimumAcceptedPoW":0,"BloomFilterMode":false,"LightClient":false,"FullNode":false,"RestrictLightClientsConn":true,"EnableConfirmations":false,"SoftBlacklistedPeerIDs":null}}
...
2023-12-24T05:37:40.729Z	DEBUG	node/node.go:185	Epoch processing	{"mvds": {"node": "04d99002", "epoch": 2252}}
fatal error: runtime: C malloc failed

goroutine 2688981 [running]:
runtime.throw({0x2b63102?, 0xc0dc0f5598?})
	/nix/store/yri46n3pjlhz9in6v7v2sz4235zz9hld-go-1.19.9/share/go/src/runtime/panic.go:1047 +0x5d fp=0xc0dc0f5570 sp=0xc0dc0f5540 pc=0x449f9d
github.com/mutecomm/go-sqlcipher/v4._cgo_cmalloc(0x44)
...
rax    0x0
rbx    0x124d2d
rcx    0x7f378dc2badc
rdx    0x6
rdi    0x124d2c
rsi    0x124d2d
rbp    0x7f376670d6c0
rsp    0x7f376670c6a0
r8     0x0
r9     0x1
r10    0x8
r11    0x246
r12    0x7f376670c970
r13    0x6
r14    0xc0000069c0
r15    0x7f3655f483a9
rip    0x7f378dc2badc
rflags 0x246
cs     0x33
fs     0x0
gs     0x0
FAIL	github.com/status-im/status-go/protocol	7598.302s
FAIL
```
https://ci.status.im/job/status-go/job/tests-nightly/72/artifact/protocol/test.log

**Comment by yakimant** [comment]
I repinned the builds for investigation.

Issue number 1 - build stuck on `status-go/api`:
https://ci.status.im/job/status-go/job/tests-nightly/71/

Issues 2 and 3 - parsing and malloc crash:
https://ci.status.im/job/status-go/job/tests-nightly/72/

@osmaczko, please have a look on those 3 issues and let me know if I can help

**Comment by yakimant** [comment]
@osmaczko, shall we close this issue?

**Comment by osmaczko** [comment]
> @osmaczko, shall we close this issue?

Hey @yakimant, sorry, I didn't have enough resources to look into that earlier. I think we can close it. I will create separate issues and take a look at them later this week.
